spec/javascripts/components/url-preview-spec.js

```diff
@@ -26,41 +26,66 @@ describe('URL preview component', function () {
-    input.dispatchEvent(new window.Event('blur'))
+  it('displays a default message when the input is empty', function () {
+    urlPreview.input.value = ''
+    urlPreview.input.dispatchEvent(new window.Event('blur'))
 
-    var defaultMessage = document.querySelector('.js-url-preview-default-message')
-    expect(defaultMessage).toBeVisible()
+    expect(urlPreview.defaultMessage).toBeVisible()
+    expect(urlPreview.urlPreview).toBeHidden()
+    expect(urlPreview.errorMessage).toBeHidden()
+  })
+
+  it('displays the preview URL when a user inputs a value and a slug is generated', function (done) {
+    fetchMock.get(/generate-path\?title=My\+title/, 'my-title')
 
-    var previewURL = document.querySelector('.js-url-preview-url')
-    expect(previewURL).toBeHidden()
+    urlPreview.input.value = 'My title'
+    urlPreview.input.dispatchEvent(new window.Event('blur'))
 
-    var errorMessage = document.querySelector('.js-url-preview-error-message')
-    expect(errorMessage).toBeHidden()
+    fetchMock.flush(true).then(function () {
+      expect(urlPreview.defaultMessage).toBeHidden()
+      expect(urlPreview.urlPreview).toBeVisible()
+      expect(urlPreview.urlPreview).toContainText('my-title')
+      expect(urlPreview.errorMessage).toBeHidden()
+      done()
+    })
   })
 
-  it('should only display the preview URL if input has value', function () {
-    // Interact with input and set value
-    var input = document.querySelector('[data-url-preview="input"]')
-    input.value = 'My title'
-    input.dispatchEvent(new window.Event('blur'))
+  it('displays an error when the request to generate a path fails', function (done) {
+    fetchMock.get(/generate-path\?title=My\+title/, 500)
+
+    urlPreview.input.value = 'My title'
+    urlPreview.input.dispatchEvent(new window.Event('blur'))
+
+    fetchMock.flush(true).then(function () {
+      expect(urlPreview.defaultMessage).toBeHidden()
+      expect(urlPreview.urlPreview).toBeHidden()
+      expect(urlPreview.errorMessage).toBeVisible()
+      done()
+    })
+  })
 
-    var defaultMessage = document.querySelector('.js-url-preview-default-message')
-    expect(defaultMessage).toBeHidden()
+  it('displays an error if the request takes longer than 5 seconds', function (done) {
+    jasmine.clock().install()
+    fetchMock.get(/generate-path\?title=My\+title/, 'my-title', { delay: 10000 })
 
-    var previewURL = document.querySelector('.js-url-preview-url')
-    expect(previewURL).toBeVisible()
+    urlPreview.input.value = 'My title'
+    urlPreview.input.dispatchEvent(new window.Event('blur'))
 
-    var errorMessage = document.querySelector('.js-url-preview-error-message')
-    expect(errorMessage).toBeHidden()
+    jasmine.clock().tick(5001)
+    fetchMock.flush(true).then(function () {
+      expect(urlPreview.defaultMessage).toBeHidden()
+      expect(urlPreview.urlPreview).toBeHidden()
+      expect(urlPreview.errorMessage).toBeVisible()
+      jasmine.clock().uninstall()
```

Comment:
I didn't know about that feature - I'll give it a go

Reply:
Yup that worked thanks - I wasn't sure if it would have problems as it'll run the install before some of our asynchronous code completes. However this doesn't seem to cause issues.